# 09.异常

## 目标

- 异常的概念
- 捕获异常
- 异常的遗传
- 自定义异常

<br><br><br>

## 1、异常的概念

- 程序在运行时，如果`Python 解释器` **遇到** 一个错误，会停止程序的执行，并且提示一些错误，叫做**异常**
- **程序停止执行并且提示错误信息** 这个动作，称之为：**抛出异常**

	```python
	# 此时，如果接受的为非数字，则程序报错，跑出异常
	num = int(input(请输入数字：))
	print(num)
	```
	
> swift `arrest()`方法？

> *程序开发时，很难将 **所有特殊的情况** 都处理的面面俱到，通过 **异常捕获** 可以针对突发事件做击中的处理，从而保证程序的 **稳定性和健壮性***

<br><br><br>

## 2、捕获异常

### 2.1 简单的捕获异常语法

- 在程序开发中，如果 **对某些代码的执行不能确定是否正确** ，可以增加`try`来 **捕获异常** 

- 捕获异常的语法

```python
try:
	尝试执行的代码
except:
	出现错误的处理
```

- `try` **尝试**，下方编写要尝试代码，不确定是否能够正常执行的代码

- `except`**如果不是**，下方编写尝试失败的代码

### 2.2 错误类型捕获

- 在程序执行时，可能会遇到 **不同类型的异常** ,并且需要针对不同类型的异常，做出不同的相应，这个时候，就需要 **捕获错误类型**

- 语法如下

```python
try:
	# 尝试执行的代码
	pass
except 错误类型1：
	# 针对错误类型1，应对的代码处理
	pass
except (错误类型2, 错误类型3)：
	# 针对错误类型2、3，应对的代码处理
	pass

```

- 当`Python`解释器 **抛出异常** 时，**最后一行错误信息的第一个单词**，就是**错误类型**

**捕获未知错误**

- 在开发时，完全预判到所有可能的错误有难度

- 所以需要捕获**未知错误**

```python
try:
	# 尝试执行的代码
	pass
except 错误类型1：
	# 针对错误类型1，应对的代码处理
	pass
except (错误类型2, 错误类型3)：
	# 针对错误类型2、3，应对的代码处理
	pass
except Exception as result:
	print("未知错误 %s " result)
```

### 2.3 异常捕获的完整语法

- 在实际开发中，为了能够处理复杂的异常情况，完整的异常语法如下：

```python
try:
	# 尝试执行的代码
	pass
except 错误类型1：
	# 针对错误类型1，应对的代码处理
	pass
except (错误类型2, 错误类型3)：
	# 针对错误类型2、3，应对的代码处理
	pass
except Exception as result:
	print("未知错误 %s " result)
else:
	# 没有异常才会执行的代码
	pass
finally:
	# 无论是不是有异常，都会执行的代码
	pass
```

<br><br><br>


## 3、异常的传递

- **异常的传递** —— 当 **函数/方法** 执行 **出现异常**，会 **将异常传递** 给函数/方法的 **调用一方**

- 如果 **传递到主程序**，仍然 **没有处理异常**，程序才会终止

> 这一层层传递让我想起了swift里面的 **可选链式调用**

> tips:
> - 在开发中，可以在**主函数**中增加 **异常捕获**
> - 而在主函数中调用的其他函数，主要出现异常，都会传递到主函数的 **异常捕获** 中
> - 这样就不需要在代码中，增加大量的 **异常捕获** ，能够保证代码的整洁

```python
def demo1():
    return int(input("请输入一个整数"))


def demo2():
    return demo1()


# 利用异常的传递性，在主程序捕获错误（虽然是在demo1()中出错的）
try:
    print(demo2())

except Exception as result:
    print("未知错误 %s " % result)
```

<br><br><br>


## 4、抛出raise异常

> 这就有点像`arrest()`函数了

### 4.1 应用场景

- 在开发中，出了 **代码执行出错** 解释器会 **抛出** 异常之外

- 还可以根据 **特殊需求** 主动 **抛出异常**

**注意**

- 当前函数 **只负责** 提示用户输入密码，如果密码长度不正确，需要其他函数进行额外的处理

- 因此，可以 **抛出异常** ，由其他需要处理的函数 **捕获异常**（重新定义一个错误类型？）

### 4.2 抛出异常

- `Python`中提供了一个`Exception`异常 **类**

- 在开发时，如果满足了 **特定业务需求时**， 希望 **抛出异常** ，可以：
	1. 创建一个`Exception`的 **对象**
	2. 使用`raise` **关键字** 抛出 **异常对象**

**示例**

- 提示用户输入密码，如果密码长度小于8位，抛出 **自定义异常**

```python
def input_password():

    # 1. 提示用户输入密码

    password = input("请输入密码")

    # 2. 判断密码长度，>=8 ，返回用于输入的密码
    if len(password) >= 8:
        return password

    # 3. 如果 <8 主动抛出异常

    print("异常")

    # 创建异常信息 - 可以使用错误信息字符串 作为参数
    error = Exception("密码长度不够")

    # 抛出异常对象
    raise error


try:
    print(input_password())
    
except Exception as result:

    print(result)
```














